File: CS/NetFramework/net48/09_GcComboBox/AutoFilter.cs
using System;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using GrapeCity.Win.Editors;

namespace InputManWin13_Demo._09_GcComboBox
{
    public partial class AutoFilter : InputManWin13_Demo.DemoBase_settings
    {
        public AutoFilter()
        {
            InitializeComponent();
            InitializeForm();
        }

        private void InitializeForm()
        {
            // ***** 初期設定
            checkBox1.CheckedChanged += new EventHandler(checkBox1_CheckedChanged);
            checkBox2.CheckedChanged += new EventHandler(checkBox2_CheckedChanged);
            gcNumber1.ValueChanged += new EventHandler(gcNumber1_ValueChanged);
            gcNumber2.ValueChanged += new EventHandler(gcNumber2_ValueChanged);
            gcNumber3.ValueChanged += new EventHandler(gcNumber3_ValueChanged);
            gcComboBox2.SelectedIndexChanged += new EventHandler(gcComboBox2_SelectedIndexChanged);
            gcComboBox3.SelectedIndexChanged += new EventHandler(gcComboBox3_SelectedIndexChanged);

            // xmlファイルからショートカット機能一覧取得
            System.String filePath = Application.StartupPath;
            this.dataSet1.ReadXml(filePath + "\\Resources\\prefecture.xml");
            gcComboBox1.DataSource = dataSet1.Tables[0];
            gcComboBox1.ListColumns[0].AutoWidth = false;

            // ***** 初期値
            checkBox1.Checked = false;
            checkBox2.Checked = false;
            gcNumber1.Value = gcComboBox1.AutoFilter.Interval;
            gcNumber2.Value = gcComboBox1.AutoFilter.MaxFilteredItems;
            gcNumber3.Value = gcComboBox1.AutoFilter.MinimumPrefixLength;
            gcComboBox2.SelectedIndex = Convert.ToInt32(gcComboBox1.AutoFilter.MatchingMode);
            gcComboBox3.SelectedIndex = Convert.ToInt32(gcComboBox1.AutoFilter.MatchingSource);

            for (int i = 1; i < gcComboBox1.ListColumns.Count; i++)
            {
                gcComboBox1.ListColumns[i].AutoWidth = true;
            }
            gcComboBox1.DropDown.Direction = DropDownDirection.BelowLeft;
        }

        private void checkBox1_CheckedChanged(object sender, EventArgs e)
        {
            // AutoFilterプロパティ
            gcComboBox1.AutoFilter.Enabled = checkBox1.Checked;
        }

        private void checkBox2_CheckedChanged(object sender, EventArgs e)
        {
            // AutoSelectプロパティ
            gcComboBox1.AutoSelect = checkBox2.Checked;
        }

        private void gcNumber1_ValueChanged(object sender, EventArgs e)
        {
            // AutoFilter.Intervalプロパティ
            if(gcNumber1.Value.HasValue)
            {
                gcComboBox1.AutoFilter.Interval = (int)gcNumber1.Value.Value;
            }
            else
            {
                gcNumber1.Value = gcNumber1.MinValue;
                gcComboBox1.AutoFilter.Interval = (int)gcNumber1.Value.Value;
            }
        }

        private void gcNumber2_ValueChanged(object sender, EventArgs e)
        {
            // AutoFilter.MaxFilteredItemsプロパティ
            if (gcNumber2.Value.HasValue)
            {
                gcComboBox1.AutoFilter.MaxFilteredItems = (int)gcNumber2.Value.Value;
            }
            else
            {
                gcNumber2.Value = gcNumber2.MinValue;
                gcComboBox1.AutoFilter.MaxFilteredItems = (int)gcNumber2.Value.Value;
            }
        }

        private void gcNumber3_ValueChanged(object sender, EventArgs e)
        {
            // AutoFilter.MinimumPrefixLengthプロパティ
            if (gcNumber3.Value.HasValue)
            {
                if (gcNumber3.Value.Value < 1)
                {
                    gcNumber3.Value = gcNumber3.MinValue;
                    gcComboBox1.AutoFilter.MinimumPrefixLength = (int)gcNumber3.Value.Value;
                }
                else
                {
                    gcComboBox1.AutoFilter.MinimumPrefixLength = (int)gcNumber3.Value.Value;
                }
            }
            else
            {
                gcNumber3.Value = gcNumber3.MinValue;
                gcComboBox1.AutoFilter.MinimumPrefixLength = (int)gcNumber3.Value.Value;
            }
        }

        private void gcComboBox2_SelectedIndexChanged(object sender, EventArgs e)
        {
            // AutoFilter.MatchingModeプロパティ
            gcComboBox1.AutoFilter.MatchingMode = (AutoCompleteMatchingMode)gcComboBox2.SelectedIndex;
        }

        private void gcComboBox3_SelectedIndexChanged(object sender, EventArgs e)
        {
            // AutoFilter.MatchingSourceプロパティ
            gcComboBox1.AutoFilter.MatchingSource = (FilterMatchingSource)gcComboBox3.SelectedIndex;
        }

        private void checkBox3_CheckedChanged(object sender, EventArgs e)
        {
            if (checkBox3.Checked)
            {
                gcComboBox1.ExtractedItemsIndicesChanged += GcComboBox1_ExtractedItemsIndicesChanged;
            }
            else
            {
                gcComboBox1.ExtractedItemsIndicesChanged -= GcComboBox1_ExtractedItemsIndicesChanged;
            }
        }

        private void GcComboBox1_ExtractedItemsIndicesChanged(object sender, EventArgs e)
        {
            var builder = new StringBuilder();
            builder.Append(gcComboBox1.Items.Count);
            builder.Append("件中");
            builder.Append(gcComboBox1.ExtractedItemsIndices.Count);
            builder.Append("件表示");
            builder.Append("\r\n表示中の項目のインデックス：");
            builder.Append(String.Join(",", gcComboBox1.ExtractedItemsIndices.ToArray()));
            
            MessageBox.Show(builder.ToString());
        }
    }
}

File: VB/NetFramework/net48/09_GcComboBox/AutoFilter.vb
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports GrapeCity.Win.Editors

Namespace _09_GcComboBox
    Partial Public Class AutoFilter
        Inherits InputManWin13_Demo.DemoBase_settings
        Public Sub New()
            InitializeComponent()
            InitializeForm()
        End Sub

        Private Sub InitializeForm()
            ' xmlファイルからショートカット機能一覧取得
            Dim filePath As System.String = Application.StartupPath
            Me.dataSet1.ReadXml(filePath & "\Resources\prefecture.xml")
            gcComboBox1.DataSource = dataSet1.Tables(0)
            gcComboBox1.ListColumns(0).AutoWidth = False

            ' ***** 初期値
            checkBox1.Checked = False
            checkBox2.Checked = False
            gcNumber1.Value = gcComboBox1.AutoFilter.Interval
            gcNumber2.Value = gcComboBox1.AutoFilter.MaxFilteredItems
            gcNumber3.Value = gcComboBox1.AutoFilter.MinimumPrefixLength
            gcComboBox2.SelectedIndex = Convert.ToInt32(gcComboBox1.AutoFilter.MatchingMode)
            gcComboBox3.SelectedIndex = Convert.ToInt32(gcComboBox1.AutoFilter.MatchingSource)

            For i As Integer = 1 To gcComboBox1.ListColumns.Count - 1
                gcComboBox1.ListColumns(i).AutoWidth = True
            Next
            gcComboBox1.DropDown.Direction = DropDownDirection.BelowLeft
        End Sub

        Private Sub CheckBox1_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles checkBox1.CheckedChanged
            ' AutoFilterプロパティ
            gcComboBox1.AutoFilter.Enabled = checkBox1.Checked
        End Sub

        Private Sub CheckBox2_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles checkBox2.CheckedChanged
            ' AutoSelectプロパティ
            gcComboBox1.AutoSelect = checkBox2.Checked
        End Sub

        Private Sub GcNumber1_ValueChanged(ByVal sender As Object, ByVal e As EventArgs) Handles gcNumber1.ValueChanged
            ' AutoFilter.Intervalプロパティ
            If gcNumber1.Value.HasValue Then
                gcComboBox1.AutoFilter.Interval = CInt(gcNumber1.Value.Value)
            Else
                gcNumber1.Value = gcNumber1.MinValue
                gcComboBox1.AutoFilter.Interval = CInt(gcNumber1.Value.Value)
            End If
        End Sub

        Private Sub GcNumber2_ValueChanged(ByVal sender As Object, ByVal e As EventArgs) Handles gcNumber2.ValueChanged
            ' AutoFilter.MaxFilteredItemsプロパティ
            If gcNumber2.Value.HasValue Then
                gcComboBox1.AutoFilter.MaxFilteredItems = CInt(gcNumber2.Value.Value)
            Else
                gcNumber2.Value = gcNumber2.MinValue
                gcComboBox1.AutoFilter.MaxFilteredItems = CInt(gcNumber2.Value.Value)
            End If
        End Sub

        Private Sub GcNumber3_ValueChanged(ByVal sender As Object, ByVal e As EventArgs) Handles gcNumber3.ValueChanged
            ' AutoFilter.MinimumPrefixLengthプロパティ
            If gcNumber3.Value.HasValue Then
                If gcNumber3.Value.Value < 1 Then
                    gcNumber3.Value = gcNumber3.MinValue
                    gcComboBox1.AutoFilter.MinimumPrefixLength = CInt(gcNumber3.Value.Value)
                Else
                    gcComboBox1.AutoFilter.MinimumPrefixLength = CInt(gcNumber3.Value.Value)
                End If
            Else
                gcNumber3.Value = gcNumber3.MinValue
                gcComboBox1.AutoFilter.MinimumPrefixLength = CInt(gcNumber3.Value.Value)
            End If
        End Sub

        Private Sub GcComboBox2_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles gcComboBox2.SelectedIndexChanged
            ' AutoFilter.MatchingModeプロパティ
            gcComboBox1.AutoFilter.MatchingMode = CType(gcComboBox2.SelectedIndex, AutoCompleteMatchingMode)
        End Sub

        Private Sub GcComboBox3_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles gcComboBox3.SelectedIndexChanged
            ' AutoFilter.MatchingSourceプロパティ
            gcComboBox1.AutoFilter.MatchingSource = CType(gcComboBox3.SelectedIndex, FilterMatchingSource)
        End Sub

        Private Sub CheckBox3_CheckedChanged(sender As Object, e As EventArgs) Handles checkBox3.CheckedChanged
            If checkBox3.Checked Then
                AddHandler gcComboBox1.ExtractedItemsIndicesChanged, AddressOf GcComboBox1_ExtractedItemsIndicesChanged
            Else
                RemoveHandler gcComboBox1.ExtractedItemsIndicesChanged, AddressOf GcComboBox1_ExtractedItemsIndicesChanged
            End If
        End Sub

        Private Sub GcComboBox1_ExtractedItemsIndicesChanged(sender As Object, e As EventArgs)
            Dim builder As StringBuilder = New StringBuilder()
            builder.Append(gcComboBox1.Items.Count)
            builder.Append("件中")
            builder.Append(gcComboBox1.ExtractedItemsIndices.Count)
            builder.Append("件表示")
            builder.Append(vbCrLf)
            builder.Append("表示中の項目のインデックス：")
            builder.Append(String.Join(",", gcComboBox1.ExtractedItemsIndices.ToArray()))

            MessageBox.Show(builder.ToString())
        End Sub
    End Class
End Namespace
